File: kbignum-java/src/main/java/com/papacekb/kbignum/BigDecimal.java
package com.papacekb.kbignum;

public class BigDecimal {

    private final java.math.BigDecimal delegate;

    private BigDecimal(java.math.BigDecimal delegate) {
        this.delegate = delegate;
    }

    public BigDecimal(String value) {
        this.delegate = new java.math.BigDecimal(value);
    }

    public BigDecimal(double value) {
        this.delegate = java.math.BigDecimal.valueOf(value);
    }

    public BigDecimal(long value) {
        this.delegate = java.math.BigDecimal.valueOf(value);
    }

    public BigDecimal add(BigDecimal n) {
        return new BigDecimal(this.delegate.add(n.delegate));
    }

    public BigDecimal subtract(BigDecimal n) {
        return new BigDecimal(this.delegate.subtract(n.delegate));
    }

    public BigDecimal multiply(BigDecimal n) {
        return new BigDecimal(this.delegate.multiply(n.delegate));
    }

    public BigDecimal divide(BigDecimal n) {
        return new BigDecimal(this.delegate.divide(n.delegate));
    }

    @Override
    public boolean equals(Object o) {
        return equalsImpl(o);
    }

    @Override
    public int hashCode() {
        return hashCodeImpl();
    }

    @Override
    public String toString() {
        return toStringImpl();
    }

    public boolean equalsImpl(Object o) {
        return o instanceof BigDecimal && this.delegate.equals(((BigDecimal)o).delegate);
    }

    public int hashCodeImpl() {
        return delegate.hashCode();
    }

    public String toStringImpl() {
        return delegate.toString();
    }

}

File: kbignum-mobile/kbignum/src/androidMain/kotlin/com/papacekb/kbignum/BigDecimal.kt
package com.papacekb.kbignum

actual class BigDecimal private constructor(private val delegate: java.math.BigDecimal) {

    actual constructor(value: String): this(java.math.BigDecimal(value))

    actual constructor(value: Double): this(java.math.BigDecimal.valueOf(value))

    actual constructor(value: Long): this(java.math.BigDecimal.valueOf(value))

    actual fun add(n: BigDecimal): BigDecimal = BigDecimal(this.delegate.add(n.delegate))

    actual fun subtract(n: BigDecimal): BigDecimal = BigDecimal(this.delegate.subtract(n.delegate))

    actual fun multiply(n: BigDecimal): BigDecimal = BigDecimal(this.delegate.multiply(n.delegate))

    actual fun divide(n: BigDecimal): BigDecimal = BigDecimal(this.delegate.divide(n.delegate))

    actual override fun equals(other: Any?): Boolean {
        return other is BigDecimal && this.delegate == other.delegate
    }

    actual override fun hashCode() = delegate.hashCode()

    actual override fun toString() = delegate.toString()
}